src/7fd10_len_b40.c

```diff
@@ -12,52 +24,149 @@ void func_800E6860(void) {
+    PartnerActionStatus* partnerActionStatus = &gPartnerActionStatus;
+    s32 actionState = gPlayerStatus.actionState;
+
+    if (!(gPlayerStatus.animFlags & PLAYER_STATUS_ANIM_FLAGS_100000) &&
+        !(gPlayerStatus.flags & PLAYER_STATUS_FLAGS_100) &&
+        !(currentButtons & (BUTTON_Z | BUTTON_R)) &&
+        (pressedButtons & (BUTTON_START | BUTTON_C_LEFT | BUTTON_C_RIGHT)) &&
+        ((gGameStatusPtr->mapShop == NULL) || !(gGameStatusPtr->mapShop->flags & 1)) &&
+        !(gOverrideFlags & GLOBAL_OVERRIDES_DISABLE_MENUS) &&
+        !is_picking_up_item())
+    {
+        if (!(gPlayerStatus.animFlags & PLAYER_STATUS_ANIM_FLAGS_8BIT_MARIO)) {
+            if (partnerActionStatus->actionState.b[0] == 0) {
+                if (!(gPlayerStatus.flags & PLAYER_STATUS_FLAGS_1000)) {
+                    if (actionState == ACTION_STATE_IDLE ||
+                        actionState == ACTION_STATE_WALK ||
+                        actionState == ACTION_STATE_RUN)
+                    {
+                        return TRUE;
+                    }
                 }
-            } else if (partnerActionStatus->actionState.b[3] == 8) {
-                if (actionState != ACTION_STATE_RIDE) {
-                    sfx_play_sound(0x21D);
-                } else {
-                    return 1;
+            } else if (partner_player_can_pause()) {
+                if (partnerActionStatus->actionState.b[3] == 6) {
+                    return actionState == ACTION_STATE_IDLE ||
+                           actionState == ACTION_STATE_WALK ||
+                           actionState == ACTION_STATE_RUN;
+                } else if (partnerActionStatus->actionState.b[3] == 9) {
+                    if (actionState == ACTION_STATE_RIDE) {
+                        gPlayerStatus.alpha2 = 0;
+                        return TRUE;
+                    }
+                } else if (partnerActionStatus->actionState.b[3] == 8) {
+                    if (actionState == ACTION_STATE_RIDE) {
+                        if (func_802BD7DC()) {
+                            return TRUE;
+                        }
+                        sfx_play_sound(SOUND_MENU_ERROR);
+                    }
+                } else if (partnerActionStatus->actionState.b[3] == 7) {
+                    sfx_play_sound(SOUND_MENU_ERROR);
                 }
             }
+        } else {
+            sfx_play_sound(SOUND_MENU_ERROR);
         }
     }
-
-    return 0;
+    return FALSE;
 }
-#else
-INCLUDE_ASM(s32, "7fd10_len_b40", func_800E6904);
-#endif
-
-INCLUDE_ASM(s32, "7fd10_len_b40", can_pause);
 
 void func_800E6B68(void) {
     D_8010CD00 = 0;
 }
 
-INCLUDE_ASM(s32, "7fd10_len_b40", setup_partner_popup);
+s32 setup_partner_popup(PopupMenu* menu) {
+    PlayerData* playerData = &gPlayerData;
+    s32 temp_a1;
+    s32* phi_t0;
+    s32 ret;
+    s32 i;
+
+    ret = 0;
+    for (i = 1; i < 12; i++) {
+        temp_a1 = D_8008EEC0[i];
```

Comment:
Presuming that `D_8008EEC0` isn't length 12?

Reply:
oopth. it iz